
---
title: "bulk_de"
output: html_document
date: "2024-01-12"
---

```{r setup}
knitr::opts_knit$set(root.dir = "~/Alyssa/SC_CHIP_HIV_Diabetes/")
library(Seurat)
library(tidyverse)
library(DESeq2)
library(tximport)
library(tximportData)
```

```{r}
system("gsutil cp gs://bicklab-main-storage/Data/vantage/10946-JVA/AppResults/AppResults/**/*.quant.genes.sf input_data/bulk_rnaseq/")
system("gsutil cp gs://bicklab-main-storage/Data/vantage/10946-JVA/AppResults/AppResults/**/*.quant.sf input_data/bulk_rnaseq/")
```

# Function for running DEseq.
```{r}
run_deseq = function(samples) {
  
  # Pull in data.
  rownames(samples) = samples$run

  files = file.path("input_data/bulk_rnaseq/", samples$run, ".quant.genes.sf", fsep = "")
  names(files) = samples$run
  
  txi = tximport(files, type="salmon", txOut = TRUE)
  dds = DESeqDataSetFromTximport(txi,
                                     colData = samples,
                                     design = ~ condition)
  
  # Filter low counts.
  smallestGroupSize = 3
  keep = rowSums(counts(dds) >= 10) >= smallestGroupSize
  dds = dds[keep,]
  
  # Set reference (control) factor.
  dds$condition = relevel(dds$condition, ref = "Mock")
  
  # Run DESeq.
  dds = DESeq(dds)
  res = results(dds, contrast=c("condition","TET2","Mock"))
  resLFC = lfcShrink(dds, coef="condition_TET2_vs_Mock", type="apeglm")
  
  # Arrange results.
  resOrdered = res[order(res$pvalue),] # not using shrunk values
  res05 = results(dds, alpha=0.05)
  summary(res05)
  sum(res05$padj < 0.05, na.rm=TRUE)
  
  return(res05)
  
}
```

# Run for each cell type.
```{r}
thp1_samples = read_tsv("input_data/bulk_rnaseq/bulk_rnaseq_samples.txt") %>%
  filter(celltype == "THP-1")

haec_samples = read_tsv("input_data/bulk_rnaseq/bulk_rnaseq_samples.txt") %>%
  filter(celltype == "HAEC")

thp1_results = run_deseq(thp1_samples)
haec_results = run_deseq(haec_samples)

```

# Convert results to gene symbols.
```{r}
library(org.Hs.eg.db)

thp1_results$symbols = mapIds(
  org.Hs.eg.db,
  keys = gsub("\\..*","", rownames(thp1_results)),
  column = 'SYMBOL',
  keytype = 'ENSEMBL')

as.data.frame(thp1_results) %>% rownames_to_column("gene") %>% arrange(padj) %>% write_tsv("tables/bulk_rnaseq_de/thp1_results.tsv")

haec_results$symbols = mapIds(
  org.Hs.eg.db,
  keys = gsub("\\..*","", rownames(haec_results)),
  column = 'SYMBOL',
  keytype = 'ENSEMBL')

as.data.frame(haec_results) %>% rownames_to_column("gene") %>% arrange(padj) %>% write_tsv("tables/bulk_rnaseq_de/haec_results.tsv")
```


```{r}
thp1_results = read_tsv("tables/bulk_rnaseq_de/thp1_results.tsv")
haec_results = read_tsv("tables/bulk_rnaseq_de/haec_results.tsv")

tem_genes = c("CCL20", "CDC42", "CTNNB1", "EZR", "ITGB2", "ICAM1", "LSP1", "MADCAM1", "MLCK", "PANX1", "RAC1B", "RHOA", "RHOG", "SRC", "SRF", "VASP", "VCAM1", "VEGF", "VWF", "CX43", "ESAM", "IQGAP2", "JAMA", "JAMB", "JAMC", "JAML", "LCK", "MMP2", "OCLN", "PECAM1", "TIAM1", "TIMP4", "CDH5", "ZO1", "ZO2", "ZO3", "CAV1", "CAV2", "CAVIN1", "PLVAP", "SNAP23", "SNAP25", "VAMP1", "VAMP2", "VAMP3", "VAMP8")

gene_labels = function(data, direction, xlim, ylim, few = FALSE) {
  sorted_data = data %>% 
    arrange(desc(`-log10_p_val`)) %>%
    filter(`-log10_p_val` > -log10(0.05)) %>%
    filter(abs(log2FoldChange) > 0.25) %>%
    head(10)
  if (few) {
    sorted_data = head(sorted_data, 3)
  }
  if (direction == "pos") {
    xlims = c(xlim - 0.5, xlim + 0.5)
  } else {
    xlims = c(-xlim - 0.5, -xlim + 0.5)
  }
  labels = geom_label_repel(data = sorted_data, mapping = aes(label = gene), size = 3, fill = NA, direction = "y", max.overlaps =  Inf, nudge_x = 0.05, segment.alpha = 0.25, force_pull = 0, xlim = xlims, ylim = c(ylim / 5, ylim), show.legend = FALSE, box.padding = 0, segment.linetype = "solid", label.size = 0.08, segment.curvature = ifelse(direction == "neg", -1e-40, 1e-40))
  return(list(labels))
}

library(ggrepel)

ggplot(thp1_results, aes(x = log2FoldChange, y = -log10(padj))) +
  geom_point() +
  geom_label_repel(data = subset(thp1_results %>% filter(padj < 0.0001 & log2FoldChange > 0)), aes(label = symbols), direction = "y", max.overlaps =  Inf, nudge_x = 0.05, segment.alpha = 0.25, force_pull = 0, xlim = 3, segment.curvature = 1e-40) +
   geom_label_repel(data = subset(thp1_results %>% filter(padj < 0.0001 & log2FoldChange < 0)), aes(label = symbols), direction = "y", max.overlaps =  Inf, nudge_x = 0.05, segment.alpha = 0.25, force_pull = 0, xlim = -4, segment.curvature = -1e-40) +
  theme_bw() +
  ggtitle("THP1s")

ggplot(haec_results, aes(x = log2FoldChange, y = -log10(padj))) +
  geom_point() +
  geom_label_repel(data = subset(haec_results %>% filter(padj < 0.000001 & log2FoldChange > 0)), aes(label = symbols), direction = "y", max.overlaps =  Inf, nudge_x = 0.05, segment.alpha = 0.25, force_pull = 0, xlim = 6, segment.curvature = 1e-40) +
   geom_label_repel(data = subset(haec_results %>% filter(padj < 0.000001 & log2FoldChange < 0)), aes(label = symbols), direction = "y", max.overlaps =  Inf, nudge_x = 0.05, segment.alpha = 0.25, force_pull = 0, xlim = -5, segment.curvature = -1e-40) +
  theme_bw() +
  ggtitle("HAECs")



```




# Pathway analysis.
```{r}
library(clusterProfiler)
library(org.Hs.eg.db)
library(ggplot2)
library(tidyverse)
library(RColorBrewer)
```


# Function for pathway analysis.
```{r}
find_pathways = function(cell_type_name, sample_genotype) {
  # Get file.
  folder = "tables/bulk_rnaseq_de/"

  input_filename = list.files(folder, cell_type_name)
  if (length(input_filename) == 0) {
    return()
  }
  
  # Load data.
  exp_data = read_tsv(paste0(folder, input_filename))

  gene_list = exp_data %>%
    dplyr::select(symbols, log2FoldChange) %>%
    dplyr::arrange(desc(log2FoldChange)) %>%
    tibble::deframe()
  
  tryCatch({  # Written with trycatch to avoid errors with empty lists.
    
    # Determine pathways.
    gse = gseGO(geneList = gene_list, keyType = "SYMBOL", OrgDb = "org.Hs.eg.db", ont = "ALL", minGSSize = 50, maxGSSize = 800, nPermSimple = 10000)
    gse_all = gseGO(geneList = gene_list, keyType = "SYMBOL", OrgDb = "org.Hs.eg.db", ont = "ALL", minGSSize = 2, maxGSSize = 800, nPermSimple = 10000)
    
    # Save files.
    gse_filename = paste0("rds_objects/gse_bulk_rnaseq/", cell_type_name, "_gse.rds")
    saveRDS(gse, gse_filename)
    
    gse_filename_all = paste0("rds_objects/gse_bulk_rnaseq/", cell_type_name, "_gse_all.rds")
    saveRDS(gse_all, gse_filename_all)

  }, error = function(cond) {
    message(cond)
    return("Error determining pathways.")
  })
}
```

```{r}
cell_types = c("haec", "thp1")
lapply(cell_types, find_pathways)
```

```{r}
save_plot = function(plot, file, legend = TRUE, width = 4, height = 4, png = FALSE) {
  if (png) { ggsave(plot = plot, filename =  paste0(file, ".png"), dpi = 600, width = ifelse(legend, width, width - 1.5), height = height, units = 'in') }
  ggsave(plot = plot, filename =  paste0(file, ".pdf"), dpi = 600, width = ifelse(legend, width, width - 1.5), height = height, units = 'in')
}


generate_pathway_plots = function(path_to_rds, title) {
  # Read in data.
  gse = readRDS(path_to_rds)
  
  tryCatch({
    # Make plots.
    split_plot = dotplot(gse, showCategory = 4, split = ".sign") +
        ggtitle(paste(title, "enriched pathways")) +
        facet_grid(.~.sign) +
        scale_color_gradient(low = brewer.pal(12, "Paired")[3], high = brewer.pal(12, "Paired")[4])
    plot = dotplot(gse) +
        ggtitle(paste(title, "enriched pathways")) +
        scale_color_gradient(low = brewer.pal(12, "Paired")[3], high = brewer.pal(12, "Paired")[4])
    
    # Make filename.
    split_filename = paste0("figures/pathway_analysis_cell_types/gsea_split_dotplot_", gsub("\\+", "", gsub(" ", "_", title)))
    
    # Save plot.
    save_plot(split_plot, split_filename, height = 4, width = 7)
  }, error = function(cond) {
    message(cond)
    return("Error rendering plot.")
  })
  
  
}
```


```{r}
generate_pathway_plots("rds_objects/gse_bulk_rnaseq/haec_gse_all.rds", "HAEC")
generate_pathway_plots("rds_objects/gse_bulk_rnaseq/thp1_gse_all.rds", "THP1")

```

