---
title: "metacells2"
output: html_document
date: "2023-09-25"
---

```{r setup}
knitr::opts_knit$set(root.dir = "~/Alyssa/Single_Cell_CHIP/")

library(tidyverse)
library(Seurat)
library(DESeq2)
library(SingleCellExperiment)
library(scuttle)
library(apeglm)
```

# Trim excessive data and clear remaining doublets.
```{r}
#diet_seurat = DietSeurat(scaled_filtered_annotated_seurat, assays = "RNA")
#saveRDS(diet_seurat, "rds_objects/diet_seurat.rds")

# need some more stringent removal of doublets for one lane
pANN_frame = read_tsv("input_data/pANN_frame.tsv") %>% as.data.frame()
rownames(pANN_frame) = pANN_frame$cell_ID
diet_seurat_trimmed = AddMetaData(diet_seurat, pANN_frame)

cells_to_remove = diet_seurat_trimmed@meta.data %>%
  filter(LANE == "7079-3") %>%
  filter(pANN >= 0.28) %>%
  pull(cell_ID)

diet_seurat_trimmed = diet_seurat_trimmed[,!colnames(diet_seurat_trimmed) %in% cells_to_remove]
saveRDS(diet_seurat_trimmed, "rds_objects/diet_seurat_trimmed.rds")

```

# Get counts.
```{r}
source("scripts/image_formatting.R")
diet_seurat_trimmed = readRDS("rds_objects/diet_seurat_trimmed.rds")
counts = diet_seurat_trimmed@assays$RNA
```

# Functions for differential expression.
```{r}
clump = function(agg_sce, label1, label2, cell_type) {
    agg_sce = agg_sce[, agg_sce$GROUP %in% c(label1, label2)] # Subset on group of interest.

    metadata = colData(agg_sce)[,c("GROUP", "metacell")]
    metadata$CHIP = factor(metadata$GROUP, levels = c(label2, label1)) # Establish to get the order we want.
    
    dds = DESeqDataSetFromMatrix(round(assay(agg_sce, "counts")), 
                    colData = metadata, 
                    design = ~ GROUP)
    
    smallest_group_sample_size = min(sum(metadata$CHIP == label1), sum(metadata$CHIP == label2))
    
    keep = rowSums(counts(dds) >= 10) >= nrow(metadata) * 0.85 # at least 85% of metacells have at least 10 transcripts
    dds = dds[keep,]
    dds = DESeq(dds)
    
    res = results(dds, contrast = c("GROUP", label1, label2))

    summary(res)
    res_ordered_wald = res[order(res$padj),] %>%
      as.data.frame() %>%
      filter(!is.na(padj)) %>%
      rownames_to_column("gene")
    
    # Remove hemoglobin and X biased genes.
    remove_list = c("DDX3Y", "HBB", "HBA1", "HBD", "HBZ", "HBG2", "HBG1", "HBG2", "CYB5R3", "HBE1", "HBM", "HBQ1", "XIST", "UTY", "USP9Y", res_ordered_wald$gene[grep("MRP", res_ordered_wald$gene)])
    
    filtered_data = res_ordered_wald %>%
      filter(!gene %in% remove_list)
  
    write_tsv(filtered_data, paste0("tables/metacells_de_cell_types/", cell_type, "_", label1, "_vs_", label2, ".tsv"))
    return(filtered_data)
}


get_cell_type_data_for_clumping = function(cell_type_name) {
  lower_cell_type_name = gsub(" ", "_", tolower(cell_type_name))
  dnmt3a_cell_type_metadata = read_csv(paste0("metacells_files/", lower_cell_type_name, "_dnmt3a_veh_metacells.csv")) %>%
    mutate(GROUP = "DNMT3A")
  tet2_cell_type_metadata = read_csv(paste0("metacells_files/", lower_cell_type_name, "_tet2_veh_metacells.csv")) %>%
    mutate(GROUP = "TET2")
  control_cell_type_metadata = read_csv(paste0("metacells_files/", lower_cell_type_name, "_none_veh_metacells.csv")) %>%
    mutate(GROUP = "none")
  cell_type_metadata = rbind(dnmt3a_cell_type_metadata, tet2_cell_type_metadata, control_cell_type_metadata) %>%
    dplyr::rename(Cell_ID = `...1`) %>%
    filter(metacell >= 0) %>% # negative metacell assignments given to outliers
    filter(Cell_ID %in% colnames(counts@counts))
    
  cell_type_counts = counts@counts[,cell_type_metadata$Cell_ID] # this syntax changes for seurat v4 and seurat v5 (counts vs counts@counts)
  rownames(cell_type_metadata) = cell_type_metadata$Cell_ID

  cell_type_sce = SingleCellExperiment(cell_type_counts, colData = cell_type_metadata)
  assayNames(cell_type_sce) = "counts"

  cell_type_agg_sce = aggregateAcrossCells(cell_type_sce, ids = colData(cell_type_sce)[,c("GROUP", "metacell")])
  return(cell_type_agg_sce)
}


```


# Run differential expression.
```{r}
run_chip_comparisons = function(cell_type_name) {

  agg_cell_type_data = get_cell_type_data_for_clumping(cell_type_name)

  tryCatch({
    tet2_comparison = clump(agg_cell_type_data, label1 = "TET2", label2 = "none", cell_type = cell_type_name) 
  }, error = function(cond) {
    message("TET2 comparison failed.")
  })
  tryCatch({
    dnmt3a_comparison = clump(agg_cell_type_data, label1 = "DNMT3A", label2 = "none", cell_type = cell_type_name) 
  }, error = function(cond) {
    message("DNMT3A comparison failed.")
  })
}

cell_types = c("CD14 Monos", "CD8+ T cells", "CD4+ T cells")
lapply(cell_types, run_chip_comparisons)

run_chip_comparisons("NK")
run_chip_comparisons("B")

```


```{r}
library(scater)
library(RColorBrewer)

make_vln_plots = function(cell_type, genotype, genotype_gene_list, use_cell_type = FALSE) {
  cell_type_agg_data = get_cell_type_data_for_clumping(cell_type)
  
  genotype_agg_sce = cell_type_agg_data[, cell_type_agg_data$GROUP %in% c(genotype, "none")] 

  genotype_gene_data = genotype_agg_sce@assays@data$counts[genotype_gene_list,] %>% t() %>% as.data.frame()
  genotype_gene_data$GROUP = colData(genotype_agg_sce)$GROUP
  genotype_gene_data$GROUP[genotype_gene_data$GROUP == "none"] = "Control"
  genotype_gene_data$GROUP = factor(genotype_gene_data$GROUP, levels = c("Control", "TET2", "DNMT3A"))
  
  render_vln_plot = function(gene) {
    vln_plot = ggplot(genotype_gene_data, aes(x = GROUP, y = get(gene), fill = GROUP)) +
      geom_violin() +
      ylab("Expression Level") +
      xlab("") +
      theme_classic() +
      theme(legend.title = element_blank(), plot.title = element_text(hjust = 0.5, face = "bold")) +
      scale_fill_manual(values = c("grey", ifelse(genotype == "TET2", brewer.pal(6, "Paired")[4], brewer.pal(6, "Paired")[6]))) +
      ggtitle(paste(ifelse(use_cell_type, cell_type, ""), gene))
    save_plot(vln_plot, paste0("figures/violins/", gsub("\\+", "", cell_type), "_", genotype, "_", gene, "_violin"), height = 3, width = 3)
    print(vln_plot)
    }
    
  lapply(genotype_gene_list, render_vln_plot)
}


make_vln_plots(cell_type = "CD14 Monos", genotype = "TET2", genotype_gene_list = c("AIF1", "CXCL3", "FCER2", "MNDA", "FN1"))
make_vln_plots("CD14 Monos", "DNMT3A", c("MNDA", "IFITM2", "ADGRE5"))

cd4_tet_genes = read_tsv("tables/metacells_de_cell_types/CD4+ T cells_TET2_vs_none.tsv") %>% head(30) %>% pull(gene)
cd4_dnmt3a_genes = read_tsv("tables/metacells_de_cell_types/CD4+ T cells_DNMT3A_vs_none.tsv") %>% head(30) %>% pull(gene)
cd8_tet_genes = read_tsv("tables/metacells_de_cell_types/CD8+ T cells_TET2_vs_none.tsv") %>% head(30) %>% pull(gene)
cd8_dnmt3a_genes = read_tsv("tables/metacells_de_cell_types/CD8+ T cells_DNMT3A_vs_none.tsv") %>% head(30) %>% pull(gene)

make_vln_plots(cell_type = "CD4+ T cells", genotype = "TET2", genotype_gene_list = cd4_tet_genes)
make_vln_plots(cell_type = "CD4+ T cells", genotype = "DNMT3A", genotype_gene_list = cd4_dnmt3a_genes)
make_vln_plots(cell_type = "CD8+ T cells", genotype = "TET2", genotype_gene_list = cd8_tet_genes)
make_vln_plots(cell_type = "CD8+ T cells", genotype = "DNMT3A", genotype_gene_list = cd8_dnmt3a_genes)


make_vln_plots(cell_type = "CD4+ T cells", genotype = "TET2", genotype_gene_list = c("GIMAP1", "GIMAP2", "GIMAP4", "GIMAP5", "GIMAP6", "GIMAP7"), use_cell_type = TRUE)
make_vln_plots(cell_type = "CD4+ T cells", genotype = "DNMT3A", genotype_gene_list = c("GIMAP1", "GIMAP2", "GIMAP4", "GIMAP5", "GIMAP6", "GIMAP7"), use_cell_type = TRUE)
make_vln_plots(cell_type = "CD8+ T cells", genotype = "TET2", genotype_gene_list = c("GIMAP1", "GIMAP2", "GIMAP4", "GIMAP5", "GIMAP6", "GIMAP7"), use_cell_type = TRUE)
make_vln_plots(cell_type = "CD8+ T cells", genotype = "DNMT3A", genotype_gene_list = c("GIMAP1", "GIMAP2", "GIMAP4", "GIMAP5", "GIMAP6", "GIMAP7"), use_cell_type = TRUE)

```

```{r}
library(scater)
library(RColorBrewer)

get_summary_stats = function(cell_type, gene_list) {
  cell_type_agg_data = get_cell_type_data_for_clumping(cell_type)
  
  gene_data = cell_type_agg_data@assays@data$counts[gene_list,] %>% t() %>% as.data.frame()
  gene_data$GROUP = colData(cell_type_agg_data)$GROUP
  gene_data$GROUP[gene_data$GROUP == "none"] = "Control"
  gene_data$GROUP = factor(gene_data$GROUP, levels = c("Control", "TET2", "DNMT3A"))
  gene_data$cell_type = cell_type

  gene_data = pivot_longer(gene_data, cols = all_of(gene_list), names_to = "Gene", values_to = "Count")

  
  return(gene_data)
}


cd4_data = get_summary_stats(cell_type = "CD4+ T cells", gene_list = c("GIMAP1", "GIMAP2", "GIMAP4", "GIMAP5", "GIMAP6", "GIMAP7"))
cd8_data = get_summary_stats(cell_type = "CD8+ T cells", gene_list = c("GIMAP1", "GIMAP2", "GIMAP4", "GIMAP5", "GIMAP6", "GIMAP7"))
cd14_data = get_summary_stats(cell_type = "CD14 Monos", gene_list = c("FN1", "FCER2", "IFITM2", "ADGRE5"))

summary_data = rbind(cd4_data, cd8_data, cd14_data) %>%
  group_by(cell_type, Gene, GROUP) %>%
  summarize(Mean = mean(Count), SD = sd(Count))
```

