---
title: "metacells2_cell_types_ccus"
output: html_document
date: "2023-10-11"
---

```{r setup}
knitr::opts_knit$set(root.dir = "~/Alyssa/Single_Cell_CHIP/")

library(tidyverse)
library(Seurat)
library(DESeq2)
library(SingleCellExperiment)
library(scuttle)
library(apeglm)
```

# Trim excessive data and clear remaining doublets.
```{r}
#diet_seurat = DietSeurat(scaled_filtered_annotated_seurat, assays = "RNA")
#saveRDS(diet_seurat, "rds_objects/diet_seurat.rds")

# need some more stringent removal of doublets for one lane
pANN_frame = read_tsv("input_data/pANN_frame.tsv") %>% as.data.frame()
rownames(pANN_frame) = pANN_frame$cell_ID
diet_seurat_trimmed = AddMetaData(diet_seurat, pANN_frame)

cells_to_remove = diet_seurat_trimmed@meta.data %>%
  filter(LANE == "7079-3") %>%
  filter(pANN >= 0.28) %>%
  pull(cell_ID)

diet_seurat_trimmed = diet_seurat_trimmed[,!colnames(diet_seurat_trimmed) %in% cells_to_remove]
saveRDS(diet_seurat_trimmed, "rds_objects/diet_seurat_trimmed.rds")

```

# Get counts.
```{r}
diet_seurat_trimmed = readRDS("rds_objects/diet_seurat_trimmed.rds")
counts = diet_seurat_trimmed@assays$RNA
```

# Functions for differential expression.
```{r}

aggregate = function(seurat_object) {
    sce = as.SingleCellExperiment(seurat_object, assay = "RNA") # convert to SCE
    agg_sce = aggregateAcrossCells(sce, ids = colData(sce)[,c("CCUS_GROUP", "metacell")]) # aggregate by metacell
    return(agg_sce)
}


clump = function(agg_sce, label1, label2, shrinkage = TRUE, cell_type) {
    agg_sce = agg_sce[, agg_sce$CCUS_GROUP %in% c(label1, label2)] # subset on CCUS_GROUP of interest

    metadata = colData(agg_sce)[,c("CCUS_GROUP", "metacell")]
    metadata$CCUS_GROUP = factor(metadata$CCUS_GROUP, levels = c(label2, label1))
    
    dds = DESeqDataSetFromMatrix(round(assay(agg_sce, "counts")), 
                    colData = metadata, 
                    design = ~ CCUS_GROUP)
    
    keep = rowSums(counts(dds) >= 10) >= nrow(metadata) * 0.85 # 10 # at least 85% of metacells that have at least 10 transcripts    
    dds = dds[keep,]
    dds = DESeq(dds)
  
    res = results(dds, contrast=c("CCUS_GROUP", label1, label2))

    summary(res)
    res_ordered_wald = res[order(res$padj),] %>%
      as.data.frame() %>%
      filter(!is.na(padj)) %>%
      rownames_to_column("gene")
    
    # Remove hemoglobin and X biased genes.
    remove_list = c("DDX3Y", "HBB", "HBA1", "HBD", "HBZ", "HBG2", "HBG1", "HBG2", "CYB5R3", "HBE1", "HBM", "HBQ1", "XIST", "UTY", "USP9Y", res_ordered_wald$gene[grep("MRP", res_ordered_wald$gene)])
    
    filtered_data = res_ordered_wald %>%
      filter(!gene %in% remove_list)
    
    
    write_tsv(filtered_data, paste0("tables/metacells_de_cell_types_filtered_ccus/", cell_type, "_", gsub(" ", "_", label1), "_vs_", gsub(" ", "_", label2), ".tsv"))
    return(res_ordered_wald)
}

get_cell_type_data_for_clumping = function(cell_type_name, condition) {
  lower_cell_type_name = gsub(" ", "_", tolower(cell_type_name))
  
  dnmt3a_chip_cell_type_metadata = read_csv(paste0("metacells_files_ccus/", lower_cell_type_name, "_dnmt3a_chip_veh_metacells.csv")) %>%
    mutate(CCUS_GROUP = "DNMT3A_CHIP_VEH")
  tet2_chip_cell_type_metadata = read_csv(paste0("metacells_files_ccus/", lower_cell_type_name, "_tet2_chip_veh_metacells.csv")) %>%
    mutate(CCUS_GROUP = "TET2_CHIP_VEH")
  tet2_ccus_cell_type_metadata = read_csv(paste0("metacells_files_ccus/", lower_cell_type_name, "_tet2_ccus_veh_metacells.csv")) %>%
    mutate(CCUS_GROUP = "TET2_CCUS_VEH")
  control_cell_type_metadata = read_csv(paste0("metacells_files_ccus/", lower_cell_type_name, "_none_control_veh_metacells.csv")) %>%
    mutate(CCUS_GROUP = "none_Control_VEH")
  
  cell_type_metadata = rbind(dnmt3a_chip_cell_type_metadata, tet2_chip_cell_type_metadata, tet2_ccus_cell_type_metadata, control_cell_type_metadata) %>%
    dplyr::rename(Cell_ID = `...1`) %>%
    filter(metacell >= 0) %>% # negative metacell assignments given to outliers
    filter(Cell_ID %in% colnames(counts@counts))
    
  cell_type_counts = counts@counts[,cell_type_metadata$Cell_ID] # this syntax changes for seurat v4 and seurat v5 (counts vs counts@counts)
  rownames(cell_type_metadata) = cell_type_metadata$Cell_ID

  cell_type_sce = SingleCellExperiment(cell_type_counts, colData = cell_type_metadata)
  assayNames(cell_type_sce) = "counts"

  cell_type_agg_sce = aggregateAcrossCells(cell_type_sce, ids = colData(cell_type_sce)[,c("CCUS_GROUP", "metacell")])
  return(cell_type_agg_sce)
}


```

```{r}
run_chip_comparisons = function(cell_type_name) {
  agg_cell_type_data = get_cell_type_data_for_clumping(cell_type_name)

  tryCatch({
    dnmt3a_chip_veh = clump(agg_cell_type_data, label1 = "DNMT3A_CHIP_VEH", label2 = "none_Control_VEH", cell_type = cell_type_name) 
  }, error = function(cond) {
    message("DNMT3A_CHIP_VEH comparison failed.")
  })
  
  tryCatch({
    tet2_ccus_veh = clump(agg_cell_type_data, label1 = "TET2_CCUS_VEH", label2 = "none_Control_VEH", cell_type = cell_type_name) 
  }, error = function(cond) {
    message("TET2_CCUS_VEH comparison failed.")
  })

  tryCatch({
    tet2_chip_veh = clump(agg_cell_type_data, label1 = "TET2_CHIP_VEH", label2 = "none_Control_VEH", cell_type = cell_type_name) 
  }, error = function(cond) {
    message("TET2_CHIP_VEH comparison failed.")
  })

}
```

```{r}
run_chip_comparisons("CD14 Monos")
run_chip_comparisons("CD8+ T cells")
run_chip_comparisons("CD4+ T cells")

```

